### ISSUE TITLE
[network] Add Rate Limiting

### ISSUE THREAD

**Issue by gregnazario** [diem/diem#6859]
# Background
## Previous rate limiting in Diem
Originally, Diem did have a rate limiting system to rate limit by number of messages.  This would lock around a semaphore counting the number of requests that had come through the system, in a confusing way that caused a memory leak.  This was removed, because it caused a memory leak, and it only gave the ability to rate limit by total number of messages (after they had been read), and not based on their content.

## Problem
We want the ability to protect nodes in Diem, both validators and full nodes, from being overloaded by bursts of expected or unexpected traffic.

### Goals
* Rate limit bandwidth inbound and outbound to a node
* Configurable - ability change limits (minimally on startup of a node)
* Keep a low overhead - allow for garbage collection to prevent secondary attacks

#  Approach
## tl;dr
I realized, this document became very long below here.  So, here's the summary:
1. We'll use [Governor](https://docs.rs/governor/0.3.1/governor/) for our rate limiting
2. We'll rate limit by IP on bytes transferred for a throughput rate limiting
3. We'll rate limit at the transport level reading in max 1 extra message, and buffering it
4. It'll be implemented as first, a dropping message solution, followed by a backpressure solution

## Rate limiting library
After taking a look at many rate limiting libraries, and looking at building my own, it looks like [Governor](https://docs.rs/governor/0.3.1/governor/) is the preferred option.

## Where do we rate limit?
There are pros and cons to where we rate limit messages.  I ideally want to keep the framework flexible enough we could have multiple layers for different purposes.  I'll go over below the options below with pros and cons, but I want to start the MVP with a Transport Layer (per message) approach.

### Transport Layer (Per byte)
As bytes are read off the socket, we can keep track of how many bytes come through for all connections on an IP or peer.  The bytes read would have to be buffered at a frame level until the rate limiter allows it through later.

Pros:
* Can provide backpressure - stop reading from socket = backpressure
* Can rate limit on a per byte level
* Cheap - Won't read a single byte more than it has to

Cons:
* Low level - can only use IP information, can't take advantage of trusted peers / protocol or message information
* Partial frame reads between throttles, can be buffered between frames

### Transport Layer (Per message) [Preferred MVP]
As message frames are read off the socket, we can keep track of how many bytes come through for all connections on an IP or peer.  This however means we may read at most one message per connection extra (if there is a thread per connection).  This message would then need to be either dropped or buffered for once it can.

Pros:
* Can provide backpressure - stop reading from socket after a message
* Can rate limit on a per byte per message level
* Simple, and will always read full frames off the socket

Cons:
* Backpressure is after reading a message too far
* Buffering a single message just to adhere to rate limit or otherwise, allow message through and go one message over
* Low level - can only use IP information, can't take advantage of trusted peers / protocol or message information 

### Application layer (Per message)
After messages are processed by the transport layer, examine message to determine if it should be rate limited.  This will allow to bring in other application level pieces like content of messages.

Pros:
* Can use protocol or message information in rate limiting decisions
* Simple, and will always read full frames off the socket

Cons:
* Can't provide backpressure, would have to drop message and let application level handle retries (DirectSend doesn't)
* High level - already incurred the cost of reading / processing message at transport layer

## What are we rate limiting on?
We'll rate limit on the IP over the PeerId, which would prevent an application or attacker from just providing a bunch of connections with different PeerIds.  This could slow multiple users with different PeerIds behind the same IP, but that's a less likely scenario.

We'll also rate limit on the number of bytes total per message, at a per second rate.  This will allow us to have a *throughput* rate per IP.  I've decided not to do anything dynamic about changing throttle sizes based on number of connections for simplicity.

## How will this be worked on?
This is all about building MVPs, so the first iteration will be to drop messages as they go over the limit to load shed.  This is problems with some of the ways our protocols are implemented (DirectSend), so we'll need to after that come back and make a backpressure solution.  However, doing the drop message system first will allow us to move faster with some side effects on pieces that use DirectSend.

**Comment by davidiw** [comment]
This is done :D
